Consolidating sky and weather effects around SkyX

ohlidalp opened the pull request:
This branch was split out from the OGRE14 port because it works just as well under current OGRE and signifficantly improves user experience:
* The top menu "Settings" is now tabbed: General/Sky&Weather/Water
* The "Sky&Weather" tab allows real time switching of the sky subsystems, including turning sky off completely (ambient light 100%).
* For SkyX, there are new sliders for daytime and moon phase.
* The precipitation effect was ported from Caelum to SkyX and can be controlled in the menubar.
* The SkyX volumetric clouds can be also adjusted realtime via sliders.
* There is a new config file in the user's profile, "config/skyx.cfg" which allows customizing all of the parameters. The format is the same as "SkyXDefault.skx" in "resources/SkyX.zip".

<img width="1067" height="527" alt="image" src="https://github.com/user-attachments/assets/4d1400d3-ca31-44f5-b8fb-86f5d769ff85" />

The motivation is to abandon Caelum because it uses Cg shaders exclusively and we want to ditch Cg (obsolete proprietary nVidia lib)..

ohlidalp committed:
:wrench: SkyX: Fixed non-UTF8 sources.

ohlidalp committed:
:video_game: Corrected camera near/far clip distances.

CVar 'gfx_sight_range' no longer determines camera far clip distance. Other uses (manual LODs and terrain fog) are unchanged.
Camera clip distances are now hardcoded <0.1, 100000.0> which works with HydraX and SkyX without needing additional adjustments.

ohlidalp committed:
:video_game: Added SkyX time-of-day slider to Top Menubar :sunny:

ohlidalp committed:
:video_game: Sky mode is now toggleable via Top Menubar UI

Other changes:
 * :envelope: Added `MSG_EDI_REINIT_SKY_REQUESTED`
 * SkyX daytime added to savegames.
 * Added 'None' sky option

ohlidalp committed:
🌧 Precipitation system ported from Caelum to SkyX

Configuration is done via SkyX script, settings have new names, see table below.

```
    // =================================================================================
    // Names in Caelum script   | names in SkyX script                       | Descriptions
    // ---------------------------------------------------------------------------------
    // "texture"                | "PrecipitationTextureName"                 | 'precipitation_X.png', where X is: 'drizzle/icecrystals/icepellets/smallhail/hail/snow/snowgrains'
    // "precipitation_colour"   | "<color>PrecipitationColor"                |
    // "falling_speed"          | "<float>PrecipitationFallingSpeed"         |
    // "wind_speed"             | "<vector3>PrecipitationWindSpeed"          | Wind speed and direction
    // "camera_speed_scale"     | "<vector3>PrecipitationCameraSpeedScale"   | Controls how much of an effect moving the camera has on rain drop directions.
    // "intensity"              | "<float>PrecipitationIntensity"            |
    // "auto_disable_intensity" | "<float>PrecipitationAutoDisableThreshold" | Auto-disable compositors when intensity is below threshold (expensive switch!). Use negative value to always keep enabled.

```

ohlidalp committed:
:moon: Added moon phase & time factor sliders to top menubar

ohlidalp committed:
:bug: Fixed SkyX skydome glitched verts

Problem: while moving camera, some verts could disappear (only changed on camera angle change), creating gap in the skydome.

Cause: Far clipping & float imprecision.

Solution: Make the skydome smaller.

ohlidalp committed:
:cloud: Added cloud sliders to TopMenubar UI

This introduces a new tab 'Weather' which contains Precipitation (previously under Sky) and newly VClouds.

ohlidalp committed:
:sunny: Fixed SkyX sun/moon trajectories

TopMenubar: sliders for sunrise/sunset time are gone, there are Lattitude and DayOfYear settings now.

ohlidalp committed:
:wrench: Refactored top menubar Settings tab into functions

ohlidalp committed:
🌦 Added weather UI to top menubar

Additional merge from the OGRE14 branch.
•	Added a “Weather (precipitation)” section for SkyX that wires directly to SkyXPrecipitationController (same API used by CfgFileManager::loadPrecipitationParams()).
•	Exposed all important runtime parameters: preset, intensity, color, falling speed, wind, camera influence, auto/manual camera speed, falling direction, auto-disable threshold, and texture (when custom).

cosmic-vole commented:
I managed to build this on my Linux box and it looks really nice. The only slight problem is it still rains and snows inside the vehicles. I did try and make some code ages ago to do RTT when inside the vehicles with Caelum but it did slow the game down a fair bit.

ohlidalp committed:
:cloud: SkyX cloud sliders now work smoothly (no rebuild lag)

ohlidalp committed:
:cloud: Added cloud coverage slider to top menubar

ohlidalp committed:
🌧 Precipitation wind speed/dir synced to vClouds

the separate config option '<vector3>PrecipitationWindSpeed' and associated XYZ widget in top menubar were removed

ohlidalp committed:
:cloud: Fixed SkyX time hotkeys and improved UI

The top menubar/settings/Sky tab was overdone, now it shows just a few sliders by default (Time of day, moonphase, cloudiness, windspeed, winddir).

Time adjustment hotkeys now control daytime directly (same as Caelum) instead of merely speeding up/slowing progression. It also now logs the "Time set to HH:MM:SS" message as Caelum does.

The default daytime progression was slowed to sane pace, circa +1min every 2 real sec. Removed param '<float>TimeMultiplier' from skyx config file.

Added '<float>Latitude' and '<int>DayOfYear' to skyx config file. Removed param '<float>TimeMultiplier'

Default wind speed reduced from 80 to 25

ohlidalp committed:
🌧 Precipitation system simplified and polished

Camera movement effect was tuned down - We only want precipitation to react to player movement, not camera orbiting/zooming. See SkyxManager.cpp.

The control panel was moved from separate 'Weather' to 'Sky & Weather' tab and pruned - by default only Preset combo and Intensity slider are shown.

ohlidalp committed:
🌧 Precipitation.cg converted to GLSL/HLSL :tada:

Zentro commented:
Going off of the build artifact, I'm seeing a crash on loading any map. Notably, it's either during or after "Loading Objects" (note, using DirectX 9):
```
22:09:46:  ===== TERRAIN LOADING DONE simple2.terrn2
22:09:46: [RoR|CVar]      sim_terrain_name:  "simple2.terrn2" (was: "")
22:09:46: [RoR|CVar]  sim_terrain_gui_name:  "Simple Test Terrain" (was: "")
22:09:46: Mesh: Loading character.mesh.
22:09:46: Skeleton: Loading character.skeleton
22:09:46: Can't assign material 'character' to SubEntity of 'Character#0_mesh' because this Material does not exist in group 'MeshesRG'. Have you forgotten to define it in a .material script?
22:09:46: Texture 'character.dds': Loading 1 faces(PF_X8R8G8B8,1024x1024x1) with 10 custom mipmaps from Image. Internal format is PF_X8R8G8B8,1024x1024x1.
22:09:46: Texture 'character-alpha.png': Loading 1 faces(PF_A8R8G8B8,1024x1024x1) Internal format is PF_A8R8G8B8,1024x1024x1.
22:09:46: Mesh: Loading arrow2.mesh.
22:09:46: Texture 'targetarrowtex.dds': Loading 1 faces(PF_DXT3,128x128x1) with 7 custom mipmaps from Image. Internal format is PF_DXT3,128x128x1.
22:09:46: [RoR|CVar]             sim_state:  "1" (was: "0")
22:09:46: [RoR|CVar]             app_state:  "2" (was: "1")
22:09:46: [RoR|General|Notice] Press CTRL+G to spawn a vehicle
22:09:46: Texture 'lightbulb.png': Loading 1 faces(PF_A8R8G8B8,16x16x1) Internal format is PF_A8R8G8B8,16x16x1.
22:09:46: Texture 'Warning': Loading 1 faces(PF_R5G6B5,8x8x1) Internal format is PF_R5G6B5,8x8x1.
22:09:52: [RoR|CVar]          gfx_sky_mode:  "3" (was: "1")
22:09:52: Warning: material Caelum/PrecipitationMaterial has no supportable Techniques and will be blank. Explanation: 
Pass 0: vertex program Caelum/PrecipitationVP cannot be used - not supported.

22:09:52: Warning: material Caelum/PrecipitationMaterial has no supportable Techniques and will be blank. Explanation: 
Pass 0: vertex program Caelum/PrecipitationVP cannot be used - not supported.

22:09:52: [Precipitation] Can't load compositor "Caelum/PrecipitationCompositor"
22:09:52: [SkyX] skyx.cfg loaded.
22:09:52: Texture 'Noise.jpg': Loading 1 faces(PF_R8G8B8,1024x1024x1) Internal format is PF_X8R8G8B8,1024x1024x1.
22:09:52: Can't assign material '_NULL_' to SubEntity of 'SkyXMeshEnt' because this Material does not exist in group 'OgreAutodetect'. Have you forgotten to define it in a .material script?
22:09:52: Texture 'SkyX_Starfield.png': Loading 1 faces(PF_A8R8G8B8,512x512x1) Internal format is PF_A8R8G8B8,512x512x1.
22:09:52: Texture 'SkyX_Moon.png': Loading 1 faces(PF_A8R8G8B8,512x512x1) Internal format is PF_A8R8G8B8,512x512x1.
22:09:52: Texture 'SkyX_MoonHalo.png': Loading 1 faces(PF_SHORT_RGBA,512x256x1) Internal format is PF_SHORT_RGBA,512x256x1.
22:09:52: CompositorChain: Compositor Caelum/PrecipitationCompositor has no supported techniques.
```

No crash on a manual build. Looks good.

ohlidalp commented:
I just ran into this myself - turns out I've had dirty build dir and things worked half by accident.
Reverting this to draft as the Precipitation port fro Cg to GLSL isn't functional here (works under OGRE14 tho).

UPDATE: Fixed, all should work now.

UPDATE2: Backing out again... something's wrong with light sources, SkyX light doesn't reflect from actors or ground. I'm using the 'OGRE inspector' gadget to figure out what's wrong.

UPDATE3: Done.

ohlidalp committed:
:scroll: Ogre inspector: Light info: added specular and derived transforms

ohlidalp committed:
:sunny: Fixed SkyX directional lights

* SkyXManager: fixed strange light setup, now using 2 directional lights (sun/moon) attached to scene nodes.
* SkyXManager: retouched gradients and sampling logic to fit with sky and provide decent ambient light.
* SkyX: added grouping scene node and associated getter. VClouds constructor now accepts `SkyX*` to allow access to the grouping scene node.

ohlidalp committed:
:broom: SkyX: tabs -> spaces

ohlidalp committed:
:sunrise: Ported Caelum's astronomy and lighting to SkyX

Summary:
A limited subset of Caelum's source files was placed in '/source/main/gfx/skyx/CaelumPort' and integrated into SkyX.

Features:
* Astronomical calculations using julian calendar
* Directional and ambient scene lighting (CAELUM_COMPONENT_SUN/MOON)
* Point starfield rendering (CAELUM_COMPONENT_POINT_STARFIELD)
* NOTE: Precipitation was ported separately earlier, see '../Precipitation'

Code changes:
* namespace changed from `Caelum` to `CaelumPort`
* Added proper grouping scene node
* Ad-hoc changes are marked with 'RIGSOFRODS:' comments

ohlidalp commented:
I've ported Caelum's astronomical and light calculations to SkyX. I haven't compared it thoroughly but it seems to work just right. There's some glitches like stars being visible at daytime and the topmenu slider being stuck at maximum. Hotkeys work OK though. I plan to also replace SkyX's sun shader colors with Caelum's.

ohlidalp committed:
:sunrise: CaelumPort: Fixed daytime slider in top menubar

ohlidalp committed:
:sunrise: CaelumPort: Fixed crash on SkyX shutdown (dupli. cleanup).

ohlidalp committed:
:sunrise: CaelumPort: converted starfield shader from Cg to HLSL/GLSL

I needed to rename the material from Caelum* to CaelumPort*, so while at it, I converted it.

ohlidalp committed:
:sunrise: CaelumPort: Fixed glitched ground lighting

OgreTerrain has a single-light limitation (it will go away after OGRE14+ port).
Worked around by hard-disabling moon light.
The old Caelum integration has moon light disabled also.

ohlidalp commented:
Converting back to draft once again because the goal shifted: originally I just wanted to port Precipitation to SkyX and fix up SkyX's own lighting, but Discord chat had additional requests:
* port lighting from Caelum, too: https://discord.com/channels/136544456244461568/189904947649708032/1489700153790300160
* make it scriptable: https://discord.com/channels/136544456244461568/189904947649708032/1491944019880575088

That turns this branch into a mess because it would add and then remove plenty of code. I'll finish the work here as research, but then I'll create a new clean one and close this PR.
